WIP: Refactor internationalization system

BertalanD opened the pull request:
This pull request implements the redesigned localization system as described in my [forum post].

In summary, the goal is to be transparent towards templates, but customizable and explicit towards users. A [Project Fluent]-based string translation system will eventually be introduced that's similar to #1040 by @XAMPPRocky.

See the commit messages for an in-depth changelog.

# To-do
- [X] Refactor existing code
- [ ] Merge `extra` from page/section header?? + language in front matter
- [x] Add proper linking in doc comments
- [ ] Add more tests
- [x] Add fluent function
- [ ] Load locales from both site and theme
- [ ] Generate Fluent stuff on init
- [ ] Reduce the amount of `clone()`s (kind of done)
- [ ] User testing
- [X] Add user documentation
- [ ] Add template API documentation
- [ ] Create a demo theme or internationalize an existing theme

[forum post]: https://zola.discourse.group/t/rfc-internationalization-system-rework/546
[Project Fluent]: https://www.projectfluent.org/

BertalanD committed:
Refactor internationalization system

This commit starts the rework of the localization system, as described
in my [forum post] a while ago. Its intent is to make it easier to
eventually transition to a [Fluent]-based localization system and allow
the official [Rust website] to move to Zola.

Monolingual sites should continue working without change. If you use the
current localization system, take a look at the summary at the end of
the commit message.

User documentation has not yet been created, so the key changes are
described here as follows:

- Require `config.default_language` and any additional
  `config.languages` to be valid Unicode language identifiers. These
   are compatible with HTML's [BCP 47] language codes and allow for
   correct language-dependent behavior (e.g. date formatting...)

- Added `language_alias` field for `config` and all languages in
  `config.languages`, which sets how the language should be searched in
  file names and displayed in URLs. This lets us avoid breaking links to
  existing multilingual sites.

- Added struct `LocaleOptions`, which contains language-specific
  settings. The title, description, taxonomies, feed limit, search
  options and `extra` keys are now also set here.

- `LocaleOptions` belonging to the default language are in the field
  `default_language_options`. Its fields are transparently flattened,
  so `config.toml`'s syntax stays the same for monolingual sites.

- Changed type of `config.languages` from `Vec<Language>` to
  `HashMap<LanguageIdentifier, LocaleOptions>.

- Removed `config.translations`. Its functionality is taken over by
  `config.languages.{lang}.extra`.

- Missing fields in the `extra` field of translations fall back to the
  values set for the default language.

- Themes can also define translations similarly to `config.toml`. All
  fields except `extra` and `languages.{lang}.extra` are ignored.

- Removed `lang` fields of taxonomies, as they are now to be defined
  under a language.

- Added `LocalizedConfig` that wraps `Config` with
  `default_language_options` replaced with a translation's settings.
  TODO: investigate if a `SerializingConfig` should be used instead.
  `config.extra` in templates now contains the values for the specific
  language. Other langugages' `extra` is best accessed via the `trans`
  function.

- Language and date parsing from filenames is done in
  `library::content::file_info`. The new `maybe_lang` and `date` fields
  are not guaranteed to be semantically correct. There is no need any
  more to externally mutate FileInfo when setting these parameters.

- Reduced the distinction of "single-language mode". Page filenames
  are parsed as "{date}{_,-}{slug}.{lang}.md" regardless of the presence
  of translations. Section indices behave the same except dates aren't
  parsed.

- Each language has its own Tera context. This requires cloning the
  config way too many times, so a more suitable and idiomatic solution
  should be sought.

- Functions taking a `lang` argument use identifiers (not aliases!;
  like `lang` in context), and they fall back to the active language
  if not specified, not the site's default language.

- Additional dependency: [unic_langid]. This is also used by Project
  Fluent, and thus by Mozilla, so it won't likely be abandoned in the
  future.

For template authors: links constructed in templates and containing
`lang` WILL break. URLs will always use the value that's in
`language_alias`. You should replace `trans()` function calls with
directly getting the value from `config.extra`, as it allows for data
other than strings and will be deprecated in the future.

For users: unless you have a multilingual site, everything will keep
working as expected. Multilingual sites WILL stop working. The new
`config.toml` layout is demonstrated in `test_site_i18n/config.toml`

[forum post]: https://zola.discourse.group/t/rfc-internationalization-system-rework/546
[Rust website]: https://www.rust-lang.org/
[BCP 47]: https://tools.ietf.org/html/bcp47
[unic_langid]: https://crates.io/crates/unic-langid

BertalanD committed:
Add doc comments to crate components/config

areille committed:
Add better errors for minifying and upgrade crate

Keats committed:
Next version

sburris0 committed:
Add Void Linux installation instructions

Darksonn committed:
Add line highlighting to code blocks (#1131)

* Add line highlighting to code blocks

* Fix highlighting of lines

Apparently every line to be highlighted is provided in one chunk.

* Add more documentation to codeblock.rs

* Turn FenceIter into an Iterator

* Move Range to fence.rs

* Add tests

Keats committed:
Merge pull request #1154 from areille/fix-minify-error

Better minify errors

Commit:
Allow site path to contain underscores (#1162)

* Allow site path to contain underscores

Fixes site.css is not being generated if any part of the path contains
underscores

* Add tests for path with underscores

BertalanD committed:
Add `fluent` function

BertalanD committed:
Fix Clippy warnings, remove unnecessary `.clone()`s

BertalanD committed:
Fix benches

BertalanD committed:
Update user-facing documentation for localization rework

thomasetter committed:
Fix a crash when --output_dir is used with serve (#1164)

Keats commented on components/config/src/config/languages.rs:
```diff
@@ -1,16 +1,95 @@
+    ///
+    /// If unset in a translation, it will fall back to how it's set for the default language.
+    /// Its default value for the primary language is `false`.
+    pub generate_feed: Option<bool>,
```
I would rather deprecate the top `generate_feed` in favour of `feed` than having `generate_feed` everywhere.

BertalanD replied:
It would feel a bit ridiculous alongside `build_search_index`. How would you feel about also renaming `build_search_index` to `searchable`?

Keats commented on components/config/src/config/languages.rs:
```diff
@@ -1,16 +1,95 @@
+    ///
+    /// If unset in a translation, it will fall back to how it's set for the default language.
+    /// Its default value for the default language is `false`.
+    pub build_search_index: Option<bool>,
```
Note: we should show a message if the language cannot be built into a search index, only a few languages are supported right now.

Keats commented on components/config/src/config/mod.rs:
```diff
@@ -100,16 +137,10 @@ pub struct Config {
-
-    /// All user params set in [extra] in the config
-    pub extra: HashMap<String, Toml>,
 }
```
I don't know about that..
Things like `compile_sass`, `minify_html` etc are not language specific. Maybe pass something like `language_config` to the template instead?

BertalanD replied:
Yep. I mainly want to have language-specific search because CJK search indices can get huge; and tight bandwidth might exclude things like feeds

Keats commented on components/config/src/config/mod.rs:
```diff
@@ -125,8 +156,26 @@ impl Config {
+        // https://stackoverflow.com/a/46766782
+        // TODO: actually print which identifier/alias is causing a problem
+        if (1..language_names.len()).any(|i| language_names[i..].contains(&language_names[i - 1])) {
+            bail!("A language code or alias should not appear twice in config.toml");
```
just use a set instead of a vec?

Keats commented on components/config/src/config/mod.rs:
```diff
@@ -262,26 +393,39 @@ impl Config {
+                lang: lang.clone(),
+                default_language_options,
+                languages,
+                ..self.clone()
```
I guess this means we clone the Config every time we render a template?

BertalanD replied:
Unfortunately, yes. I'm thinking about having a struct with just references like `SerializingTaxonomy` and co. I'm way too inexperienced to not resort to cloning something with that many borrows.

Keats commented on components/config/src/config/mod.rs:
```diff
@@ -312,35 +454,63 @@ impl Default for Config {
-            languages: Vec::new(),
-            generate_feed: false,
-            feed_limit: None,
+            lang: langid!("en"),
```
do we need that macro? Is the macro-less approach harder to use?

BertalanD replied:
We could have it be parsed from a str. Parsing and using an `unwrap()` in a constructor seems a bit dodgy to me.

Keats commented on components/config/src/config/taxonomies.rs:
```diff
@@ -11,9 +13,12 @@ pub struct Taxonomy {
+
+    // Only for convenience/backwards compatibility
+    #[serde(skip_deserializing)]
+    pub lang: LanguageIdentifier,
```
that's still going to break if someone used an invalid language identifier right?

BertalanD replied:
This field [won't be read](https://serde.rs/field-attrs.html#skip_deserializing) from the config, just passed to templates. We set it from the containing `LocaleOptions`' key in `config.languages`. Any invalid language will error out when parsing the config.

I might not get what you're saying though.

Keats commented on components/config/src/theme.rs:
```diff
@@ -2,40 +2,55 @@ use std::collections::HashMap;
-            bail!("Expected the `theme.toml` to be a TOML table")
-        }
-
-        Ok(Theme { extra })
```
how come we don't need that anymore?

BertalanD replied:
The code previously parsed the toml into  a Toml object and just copied the `extra` key from it into the struct. This did not use Serde at all.

The new code just uses Serde to directly parse that into the `Theme` struct. This actually saves a `clone()` and having to do manual error handling. Also this code won't need to be modified for an eventual YAML or JSON (ew) config option.

I should probably look up if an attribute needs to be set to ignore unknown keys.

Keats commented on components/library/src/content/file_info.rs:
```diff
@@ -26,15 +34,48 @@ pub fn find_content_components<P: AsRef<Path>>(path: P) -> Vec<String> {
+}
+
+/// Strip `{date}{_,-}` prefix and `.{lang}` suffix from name
+pub fn clear_filename<S: AsRef<str>>(name: S) -> String {
```
I don't like the name but I can't think of anything better gith now.

BertalanD replied:
`strip` as in my comment?

Keats commented on components/library/src/content/file_info.rs:
```diff
@@ -49,26 +90,39 @@ pub struct FileInfo {
     pub canonical: PathBuf,
+    /// The `lang` part of the {name}.{lang}.md naming scheme. There is no guarantee that this
+    /// language is enabled for the site.
+    pub maybe_lang: Option<String>,
```
let's call it lang rather than maybe_lang.

Keats commented on components/library/src/content/page.rs:
```diff
@@ -110,27 +104,47 @@ impl Page {
+                // TODO: add test
+                bail!("Page `{}` has an empty language. Did you mistakenly put 2 dots before the extension?", file_path.display());
+            } else if l == &config.default_language_options.language_alias {
+                bail!("Pages in the default language should not set it in their title (caused by `{}`).", file_path.display());
```
does that cause any issues?

BertalanD replied:
Don't think so. The code currently does fail though if the default language is set in a file name.

```rust
if !config.languages_codes().contains(&parts[1].as_ref()) {
            bail!("File {:?} has a language code of {} which isn't present in the config.toml `languages`", self.path, parts[1]);
        }
```

Keats commented on components/library/src/content/page.rs:
```diff
@@ -110,27 +104,47 @@ impl Page {
+        if let Some(ref date) = page.file.date {
             if page.meta.date.is_none() {
-                page.meta.date = Some(caps.name("datetime").unwrap().as_str().to_string());
+                // TODO: do not fail silently on invalid formats
```
do we have a test for that?

Keats commented on components/library/src/content/section.rs:
```diff
@@ -86,21 +90,54 @@ impl Section {
+        // Catch programming errors
+        // `lang` and `language_alias` must be non-empty
+        assert_ne!(section.lang, LanguageIdentifier::default());
+        assert!(!section.language_alias.is_empty());
```
looks like an almost copy/paste than the page.rs file

Keats commented on components/search/src/lib.rs:
```diff
@@ -26,42 +26,45 @@ lazy_static! {
+fn build_fields(config: &LocalizedConfig) -> Vec<String> {
     let mut fields = vec![];
-    if config.search.include_title {
+    let search = config.0.default_language_options.search.clone().expect("search set");
```
can we ever have None?

Keats commented on components/search/src/lib.rs:
```diff
@@ -80,26 +83,28 @@ fn fill_index(
-    };
+pub fn build_index(library: &Library, config: &LocalizedConfig) -> Result<String> {
+    let language = Language::from_code(config.0.lang.language.as_str()).ok_or_else(|| {
+        format!("Tried to build search index for language {} which is not supported", config.0.lang)
```
we can move that error at start time

Keats reviewed (changes requested):
Ok I started reviewing it but the PR is way too big to be properly reviewed on its own.
Can we split it in smaller PRs?

BertalanD commented on components/rendering/tests/codeblock_hl_lines.rs:
```diff
@@ -40,20 +43,28 @@ fn hl_lines_simple() {
+    "#,
+        &context,
+    )
+    .unwrap();
```
Rustfmt really messed up this part. Sorry about that. I probably should revert changes to this file.

BertalanD committed:
Simplify duplicate language detection

BertalanD committed:
Tweak filename handling

- renamed `clear_filename` to `strip_filename_info`
- removed unnecessary String allocations from `get_language` and
  `get_date`

timvisee committed:
Fix #1165, improve situation for compile_sass errors (#1166)

* Update compile_sass docs, this is just for sass files in site root

* Improve file watch error, question whether the target exists
